tsl/src/compression/compression.c

```diff
@@ -323,14 +346,142 @@ compress_chunk(Oid in_table, Oid out_table, const ColumnCompressionInfo **column
+	 * orderby_asc and null_first.
+	 *
+	 * BTREE Indexes Ordering
+	 * =====================
+	 * a) ASC[Null_Last] ==> [1]->[2]->NULL
+	 * b) [Null_First]ASC ==> NULL->[1]->[2]
+	 * c) DSC[Null_Last]  ==> [2]->[1]->NULL
+	 * d) [Null_First]DSC ==> NULL->[2]->[1]
+	 */
+	if (ts_guc_enable_compression_indexscan)
+	{
+		foreach (lc, in_rel_index_oids)
+		{
+			Oid index_oid = lfirst_oid(lc);
+			Relation index_rel = index_open(index_oid, AccessShareLock);
+			IndexInfo *index_info = BuildIndexInfo(index_rel);
+			int previous_direction = NoMovementScanDirection;
+			int current_direction = NoMovementScanDirection;
 
-	Tuplesortstate *sorted_rel = compress_chunk_sort_relation(in_rel, n_keys, keys);
+			if (n_keys <= index_info->ii_NumIndexKeyAttrs && index_info->ii_Am == BTREE_AM_OID)
+			{
+				for (i = 0; i < n_keys; i++)
+				{
+					int16 att_num = get_attnum(in_table, NameStr(keys[i]->attname));
+
+					int16 option = index_rel->rd_indoption[i];
+					bool index_orderby_asc = ((option & INDOPTION_DESC) == 0);
+					bool index_null_first = ((option & INDOPTION_NULLS_FIRST) != 0);
+					bool is_orderby_asc =
+						COMPRESSIONCOL_IS_SEGMENT_BY(keys[i]) ? true : keys[i]->orderby_asc;
+					bool is_null_first =
+						COMPRESSIONCOL_IS_SEGMENT_BY(keys[i]) ? false : keys[i]->orderby_nullsfirst;
+
+					if (att_num > 0 && index_info->ii_IndexAttrNumbers[i] == att_num)
+					{
+						in_table_tp = SearchSysCacheAttNum(in_table, att_num);
+						if (!HeapTupleIsValid(in_table_tp))
+							elog(ERROR,
+								 "table \"%s\" does not have column \"%s\"",
+								 get_rel_name(in_table),
+								 NameStr(keys[i]->attname));
+
+						index_tp = SearchSysCacheAttNum(index_oid, i + 1);
+						if (!HeapTupleIsValid(index_tp))
+							elog(ERROR,
+								 "index \"%s\" does not have column \"%s\"",
+								 get_rel_name(index_oid),
+								 NameStr(keys[i]->attname));
+
+						in_table_attr_tp = (Form_pg_attribute) GETSTRUCT(in_table_tp);
+						index_attr_tp = (Form_pg_attribute) GETSTRUCT(index_tp);
+
+						if (index_orderby_asc == is_orderby_asc &&
+							index_null_first == is_null_first &&
+							in_table_attr_tp->attcollation == index_attr_tp->attcollation)
+						{
+							current_direction = ForwardScanDirection;
+						}
+						else if (index_orderby_asc != is_orderby_asc &&
+								 index_null_first != is_null_first &&
+								 in_table_attr_tp->attcollation == index_attr_tp->attcollation)
+						{
+							current_direction = BackwardScanDirection;
+						}
+						else
+						{
+							current_direction = NoMovementScanDirection;
+							break;
+						}
+
+						ReleaseSysCache(in_table_tp);
+						in_table_tp = NULL;
+						ReleaseSysCache(index_tp);
+						index_tp = NULL;
+						if (previous_direction == NoMovementScanDirection)
+						{
+							previous_direction = current_direction;
+						}
+						else if (previous_direction != current_direction)
+						{
+							break;
+						}
+					}
+					else
+					{
+						break;
+					}
```

Comment:
You're not resetting `current_direction` here, this will lead to an index still being chosen if the last column didn't match, right?

We have to fix some things:
1) add a test for this bug. Try to reproduce it on the unfixed version and verify that the last column is chosen incorrectly.
2) Reset `current_direction` at the start of the loop to avoid bugs like this.
3) Rewrite this as early exit from the loop:
```
					if (att_num == 0 || index_info->ii_IndexAttrNumbers[i] != att_num)
					{
					    break;
					}
```
Long if branches are hard to read, so early exit with break/return is preferrable.